server/src/main/java/org/elasticsearch/search/vectors/RescoreKnnVectorQuery.java

```diff
@@ -0,0 +1,160 @@
+
+        final DoubleValuesSource valueSource;
+        if (byteTarget != null) {
+            valueSource = new VectorSimilarityByteValueSource(fieldName, byteTarget, vectorSimilarityFunction);
+        } else {
+            valueSource = new VectorSimilarityFloatValueSource(fieldName, floatTarget, vectorSimilarityFunction);
+        }
+        FunctionScoreQuery functionScoreQuery = new FunctionScoreQuery(innerQuery, valueSource);
+        Query query = searcher.rewrite(functionScoreQuery);
+
+        if (k == null) {
+            // No need to calculate top k - let the request size limit the results
+            return query;
+        }
+
+        // Retrieve top k documents from the rescored query
+        TopDocs topDocs = searcher.search(query, k);
+        ScoreDoc[] scoreDocs = topDocs.scoreDocs;
+        int[] docIds = new int[scoreDocs.length];
+        float[] scores = new float[scoreDocs.length];
+        for (int i = 0; i < scoreDocs.length; i++) {
+            docIds[i] = scoreDocs[i].doc;
+            scores[i] = scoreDocs[i].score;
+        }
+
+        vectorOpsCount = scoreDocs.length;
+
+        return new KnnScoreDocQuery(docIds, scores, searcher.getIndexReader());
+    }
+
+    public Query innerQuery() {
+        return innerQuery;
+    }
+
+    public Integer k() {
+        return k;
+    }
+
+    @Override
+    public void profile(QueryProfiler queryProfiler) {
+        if (innerQuery instanceof ProfilingQuery profilingQuery) {
+            profilingQuery.profile(queryProfiler);
+        }
+        queryProfiler.addVectorOpsCount(vectorOpsCount);
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+        if (visitor.acceptField(fieldName)) {
+            visitor.visitLeaf(this);
+        }
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        RescoreKnnVectorQuery that = (RescoreKnnVectorQuery) o;
+        return Objects.equals(fieldName, that.fieldName)
+            && Objects.deepEquals(byteTarget, that.byteTarget)
+            && Objects.deepEquals(floatTarget, that.floatTarget)
+            && vectorSimilarityFunction == that.vectorSimilarityFunction
+            && Objects.equals(k, that.k)
+            && Objects.equals(innerQuery, that.innerQuery);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(fieldName, Arrays.hashCode(byteTarget), Arrays.hashCode(floatTarget), vectorSimilarityFunction, k, innerQuery);
+    }
+
+    @Override
+    public String toString(String field) {
+        final StringBuilder sb = new StringBuilder("KnnRescoreVectorQuery{");
+        sb.append("fieldName='").append(fieldName).append('\'');
+        if (byteTarget != null) {
+            sb.append(", byteTarget=").append(Arrays.toString(byteTarget));
+        } else {
+            sb.append(", floatTarget=").append(Arrays.toString(floatTarget));
```

Comment:
serializing the entire array here is VERY expensive. See how its done in other examples where we only serialize the first `n` or whatever elements.